src/plugins/git/porcelain.js

```diff
@@ -0,0 +1,263 @@
+ * Creating the wrappers is extremely cheap; all actual computation (e.g.
+ * finding the body or author of a post) is done lazily when that information
+ * is requested.
+ *
+ * The porcelain system is under active development.
+ * I expect that we will soon refactor the base porcelain abstraction out of
+ * this module into core/, and the implementation may shift to be more conserved
+ * with the GitHub porcelain. The APIs should remain unchanged.
+ * */
+import stringify from "json-stable-stringify";
+
+import {Graph} from "../../core/graph";
+import type {Node} from "../../core/graph";
+import type {Address} from "../../core/address";
+
+import type {
+  TreeEntryNodePayload,
+  SubmoduleCommitPayload,
+  NodePayload,
+  NodeType,
+  Hash,
+} from "./types";
+import {GIT_PLUGIN_NAME} from "./types";
+import {commitAddress} from "./address";
+
+export class PorcelainGraph {
+  graph: Graph<any, any>;
+  constructor(graph: Graph<any, any>) {
+    this.graph = graph;
+  }
+
+  // Note that this method is presently unsafe, as the hash may not exist.
+  // In the future, we will come up with a general case solution to have
+  // the type system verify that returned porcelains must be existence-tested
+  // before their properties are usable.
+  commitByHash(h: Hash): Commit {
+    const addr = commitAddress(h);
+    return new Commit(this.graph, addr);
+  }
+}
+
+export type GitPorcelain = Commit | Blob | Tree | TreeEntry | SubmoduleCommit;
+
+function asGitPorcelain(n: BaseNodePorcelain): GitPorcelain {
+  const type = n.type();
+  switch (type) {
+    case "COMMIT":
+      return new Commit(n.graph, n.nodeAddress);
+    case "BLOB":
+      return new Blob(n.graph, n.nodeAddress);
+    case "TREE_ENTRY":
+      return new TreeEntry(n.graph, n.nodeAddress);
+    case "TREE":
+      return new Tree(n.graph, n.nodeAddress);
+    case "SUBMODULE_COMMIT":
+      return new SubmoduleCommit(n.graph, n.nodeAddress);
+    default:
+      // eslint-disable-next-line no-unused-expressions
+      (type: empty);
+      throw new Error(
+        `Tried to make Git porcelain, but got invalid type: ${stringify(
+          n.address()
+        )}`
+      );
+  }
+}
+
+export class BaseNodePorcelain {
+  graph: Graph<any, any>;
+  nodeAddress: Address;
+
+  constructor(graph: Graph<any, any>, nodeAddress: Address) {
+    if (nodeAddress.pluginName !== GIT_PLUGIN_NAME) {
+      throw new Error(
+        `Tried to create Git porcelain for node from plugin: ${
+          nodeAddress.pluginName
+        }`
+      );
+    }
+    this.graph = graph;
+    this.nodeAddress = nodeAddress;
+  }
+
+  neighbors(options?: {|
+    +nodeType?: string,
+    +edgeType?: string,
+    +direction?: "IN" | "OUT",
+  |}): BaseNodePorcelain[] {
+    return this.graph
+      .neighborhood(this.nodeAddress, options)
+      .map(({neighbor}) => new BaseNodePorcelain(this.graph, neighbor))
+      .map(asGitPorcelain);
```

Comment:
This shouldn't be here—it will crash when a Git node has incident edges to/from other plugins' nodes (e.g., a pull-request-to-merged-commit edge).